Enable parameterized gates in C

Cryoris opened the pull request:
<!--
⚠️ If you do not respect this template, your pull request will be closed.
⚠️ Your pull request title should be short detailed and understandable for all.
⚠️ Also, please add a release note file using reno if the change needs to be
  documented in the release notes.
⚠️ If your pull request fixes an open issue, please link to the issue.

- [ ] I have added the tests to cover my changes.
- [ ] I have updated the documentation accordingly.
- [ ] I have read the CONTRIBUTING document.
-->

### Summary

This PR allows adding parameterized gates to circuit via the C API. E.g. one can do
```c
QkCircuit *qc = qk_circuit_new(2, 0);
QkParam *x = qk_param_new_symbol("x");
QkParam *y = qk_param_new_symbol("y");

uint32_t q0[1] = {0};
const QkParam *rx_param[1] = {x};
const QkParam *ry_param[1] = {y};

qk_circuit_gate_param(qc, QkGate_RX, q0, rx_param);
qk_circuit_gate_param(qc, QkGate_RY, q0, ry_param);

// check the number of parameters
size_t num_params = qk_circuit_num_free_parameters(qc);  // == 2

qk_param_free(x);
qk_param_free(y);
qk_circuit_free(qc);
```
This implies promoting the `QkInstruction.params` type from `float*` to `QkParam**`, reflecting that parameters are internally stored via `QkParam` which allows for parameterized expressions.

Fixes #14990 

### Details and comments

I'm not super happy about the function name `qk_circuit_gate_param`, open to any other suggestions (including making this `qk_circuit_gate` and renaming the existing one to `qk_circuit_gate_fixed`, like we do in the `QkTarget`).

### Ownership of `Param`

Allocations:
* The C user allocates ``QkParam *`` owned by Rust to add to the circuit.
* Upon calling `qk_circuit_gate_param`, we clone the `Param` objects and the `CircuitData` (aka `QkCircuit`) owns the new objects.
* `QkInstruction` points to an array of `QkParam *`, but does not clone or own any parameter object.

Free-ing responsibility:
* The user must call `qk_param_free` for any `QkParam *` they allocated, independently of whether these parameters are passed into a circuit or not.
* When calling `qk_circuit_free` all `QkParam` in the circuit are freed -- it is therefore _invalid_ to access ``QkInstruction.params`` after the circuit has been freed. _Note: this could be changed, `QkInstruction` could own a clone._

Cryoris committed:
Add parameterized gates

Cryoris committed:
fix C tests

jakelishman commented on crates/cext/src/circuit.rs:
```diff
@@ -351,6 +351,47 @@ pub unsafe extern "C" fn qk_circuit_num_clbits(circuit: *const CircuitData) -> u
+/// qk_circuit_gate_param(qc, QkGate_RY, q0, ry_param);
+///
+/// // check the number of parameters
+/// size_t num_params = qk_circuit_num_free_parameters(qc);  // == 2
```
This naming doesn't match the function

jakelishman commented on crates/cext/src/circuit.rs:
```diff
@@ -351,6 +351,47 @@ pub unsafe extern "C" fn qk_circuit_num_clbits(circuit: *const CircuitData) -> u
+/// Behavior is undefined if ``circuit`` is not a valid, non-null pointer to a ``QkCircuit``.
+#[unsafe(no_mangle)]
+#[cfg(feature = "cbinding")]
+pub unsafe extern "C" fn qk_circuit_num_free_params(circuit: *const CircuitData) -> usize {
```
Is the word `free` in here doing anything?  We don't use it in Python (`QuantumCircuit.num_parameters`).  Maybe we can just drop it?

Cryoris replied:
I was struggling here: in Python we associate the term `parameter` with an unbound symbol and `param` with a gate parameter. So the terminology `num_parameters` and `num_params` is well-defined (albeit confusing). In C we don't really have the term `parameter` associated the same way, so I was looking for a clearer terminology... 

Thinking about this now, why don't we call it `qk_circuit_num_symbols`? That would also match the terminology in `qk_param_new_symbol` 🙂

jakelishman commented on crates/cext/src/circuit.rs:
```diff
@@ -471,6 +512,90 @@ pub unsafe extern "C" fn qk_circuit_gate(
 }
 
+/// @ingroup QkCircuit
+/// Append a possibly parameterized ``QkGate`` to the circuit.
```
imo this isn't clear about what the difference is from `qk_circuit_gate` - I would say that `RZ` is "parametrised", even if the parameter is an `f64` and I can use `qk_circuit_gate`.  I think the same goes for the function name too - I think we're overloading `param`.

I don't have hugely better suggestions in naming, but I think we do need them.

Cryoris replied:
Yeah agreed I'm not very happy about `qk_circuit_gate_param` -- if we would name them anew it would be nice to have
```c
qk_circuit_gate(..., QkParam**, ...);
qk_circuit_gate_fixed(..., float*, ...);
```
which doesn't get into the typing or parameterized vs. parameterized and matches the ``QkTarget`` notation

jakelishman commented on crates/cext/src/circuit.rs:
```diff
@@ -908,11 +1034,11 @@ pub unsafe extern "C" fn qk_circuit_get_instruction(
-                Param::Float(val) => *val,
-                _ => panic!("Invalid parameter on instruction"),
+                Param::Obj(_) => panic!("Invalid parameter on instruction"),
+                _ => x as *const Param,
```
Just as a minor point of style/safety, it might be nice to keep the precise matching for the happy paths and the wildcard for the failure, so we default safe in the face of `Param` changing, instead of defaulting to returning something that might cause `unsafe` problems in the C interface code?

jakelishman commented on test/c/test_circuit.c:
```diff
@@ -915,6 +916,107 @@ static int test_not_unitary_gate(void) {
+    QkComplex64 matrix[16] = {c1, c1, c0, c0,  // this
+                              c1, c1, c0, c0,  // is
+                              c0, c0, c1, c0,  // for
+                              c0, c0, c0, c1}; // formatting
```
```suggestion
    // clang-format off
    QkComplex64 matrix[16] = {c1, c1, c0, c0,
                              c1, c1, c0, c0,
                              c0, c0, c1, c0,
                              c0, c0, c0, c1};
    // clang-format on
```

jakelishman commented on test/c/test_circuit.c:
```diff
@@ -915,6 +916,107 @@ static int test_not_unitary_gate(void) {
+
+    int result = Ok;
+
+    if (qk_circuit_gate(qc, QkGate_SX, q0, NULL) != QkExitCode_Success) {
+        result = RuntimeError;
+        goto cleanup;
+    }
+    if (qk_circuit_gate(qc, QkGate_RX, q0, angle) != QkExitCode_Success) {
+        result = RuntimeError;
+        goto cleanup;
+    }
+    if (qk_circuit_gate_param(qc, QkGate_R, q0, angles) != QkExitCode_Success) {
+        result = RuntimeError;
+        goto cleanup;
+    }
+    if (qk_circuit_unitary(qc, matrix, q0, 1, false) != QkExitCode_Success) {
+        result = RuntimeError;
+        goto cleanup;
+    }
```
These will attempt to free an uninitialised `inst` on failure.

jakelishman commented on test/c/test_circuit.c:
```diff
@@ -915,6 +916,107 @@ static int test_not_unitary_gate(void) {
+    // SX has no parameters
+    qk_circuit_get_instruction(qc, 0, &inst);
+    if (inst.num_params != 0) {
+        printf("Expected 0 parameters in SX, got %u", inst.num_params);
```
Missing newlines in the print statements (here and everywhere).

jakelishman commented on test/c/test_circuit.c:
```diff
@@ -915,6 +916,107 @@ static int test_not_unitary_gate(void) {
+        printf("Unexpected parameter value in RX gate");
+        result = EqualityError;
+        goto cleanup;
+    }
```
If `rz_angle` is NaN because the parameter wasn't an `f64`, this will spuriously pass because `NaN` fails all arithmetic comparisons.

Shouldn't this equality be floating-point exact, anyway?

(Same comment in a couple of places.)

jakelishman commented on test/c/test_circuit.c:
```diff
@@ -915,6 +916,107 @@ static int test_not_unitary_gate(void) {
+        goto cleanup;
+    }
+
+    // RX has one parameter
+    qk_circuit_get_instruction(qc, 1, &inst);
```
Iirc, this would overwrite the pointers in `inst` without freeing them?  (Comment repeats.)

jakelishman commented on test/c/test_circuit.c:
```diff
@@ -932,6 +1034,54 @@ static int test_delay_instruction(void) {
+    size_t num_params = qk_circuit_num_free_params(qc);
+    if (num_params != 2) {
+        result = EqualityError;
+        printf("Expected 2 free parameters, found %zu", num_params);
```
Newline

jakelishman commented on crates/cext/src/circuit.rs:
```diff
@@ -995,7 +1121,7 @@ pub unsafe extern "C" fn qk_circuit_instruction_clear(inst: *mut CInstruction) {
         inst.num_clbits = 0;
         if inst.num_params > 0 && !inst.params.is_null() {
             let params = std::slice::from_raw_parts_mut(inst.params, inst.num_params as usize);
-            let _ = Box::from_raw(params as *mut [f64]);
+            let _ = Box::from_raw(params);
             inst.params = std::ptr::null_mut();
```
This doesn't drop the `Param` objects, just the pointers to them.  It worked for `f64` because it doesn't have a `Drop` impl, but `Param` does and needs to be dropped correctly.

Cryoris replied:
On this; `qk_circuit_instruction_clear` shouldn't drop the `Param` objects, should it? `QkInstruction` holds a `Vec<*const Param>` but doesn't actually allocate or own the `Param`s, these are owned by the `CircuitData` we query the instruction from. So we only have to release the vector (i.e. the pointer to the pointers), right?

Cryoris replied:
Edit: [d8c991b](https://github.com/Qiskit/qiskit/pull/15106/commits/d8c991b0531f033b7929b5447c1006f545b8e3b7) updates the code to store copies to `QkParam` objects, in consistency with the rest of the `QkCircuitInstruction` object. I'd appreciate a detailed look on the pointer logic here though 🙂

jakelishman reviewed (commented):
In theory the API looks to me.  I think we could do some better framing with the naming, and I don't love the complexity that'll now come in `CInstruction->params`, but there's not too much to be done about that - it just is the way that Qiskit is.

Cryoris committed:
review comments Jake

Cryoris committed:
add reno

kevinhartman commented on crates/cext/src/circuit.rs:
```diff
@@ -471,6 +512,90 @@ pub unsafe extern "C" fn qk_circuit_gate(
+/// ``QkParam``.
+#[unsafe(no_mangle)]
+#[cfg(feature = "cbinding")]
+pub unsafe extern "C" fn qk_circuit_gate_param(
```
Mentioned offline, but I'd prefer something like `qk_circuit_generic_gate` or `qk_circuit_parameterized_gate`, the important part being that we end with a noun of what is being added, since that's what we're doing for everything else in `QkCircuit`. We aren't adding a "gate param" to the circuit, we're adding a "parameterized gate".

kevinhartman commented on crates/cext/src/circuit.rs:
```diff
@@ -843,7 +968,8 @@ pub struct CInstruction {
     /// A pointer to an array of parameter values for this instruction.
-    params: *mut f64,
+    /// Since ``QkParam`` are immutable this is a pointer to ``const QkParam*``.
+    params: *mut *const Param,
```
To ensure proper typing (since we currently have the problem with parameters in Rust and Python that they can be _anything_ through our interface, when in reality their type should be dictated by the operation), we might want to consider getting rid of `CInstruction::params` altogether and instead having specialized functions to query the parameters, per operation kind. For example:

- `qk_circuit_get_gate_params(...) -> *f64`
- `qk_circuit_get_generic_gate_params(...) -> *Param`
- `qk_circuit_get_control_flow_blocks(...) -> ???`
- ... and so on

We'd also need to add an enum field to `CInstruction` to tell the user what kind of instruction they're dealing with, and then it'd be up to them to call the proper accessor method(s) above to get their properly typed data.

jakelishman commented:
Decided in the team meeting today that we would defer this to 2.4, despite it originally being one of the core goals of 2.3.  That's unfortunate, but we have design decisions we didn't manage to make, given the demands on the rest of the team over the release cycle.

Cryoris committed:
Merge branch 'main' into qk-gate-param

Cryoris committed:
fix cbindgen

Cryoris committed:
Fix namings

Cryoris commented:
Not sure why but somehow GitHub doesn't show the latest commit on the [Cryoris:qk-gate-param](https://github.com/Cryoris/qiskit-terra/tree/qk-gate-param) branch where the merge conflicts are resolved 🤔

Cryoris committed:
Trying to retrigger GitHub

Cryoris committed:
Store pointers to parameter copies

QkCircuitInstruction stores copies of the PackedInstruction data, not references to it. This also holds for the parameters.

jakelishman commented on crates/cext/src/circuit.rs:
```diff
@@ -338,6 +338,46 @@ pub unsafe extern "C" fn qk_circuit_num_clbits(circuit: *const CircuitData) -> u
+///
+/// Behavior is undefined if ``circuit`` is not a valid, non-null pointer to a ``QkCircuit``.
+#[unsafe(no_mangle)]
+pub unsafe extern "C" fn qk_circuit_num_symbols(circuit: *const CircuitData) -> usize {
```
Nit: can we call it `qk_circuit_num_param_symbols` for symmetry with `qk_param_symbol_new`?  The symmetry might help ease the cognitive load a little, and "symbol" in IR terms means a lot more than just the `Parameter` stuff (e.g. register names, gate names, vars, etc).

Can we also update the docs of this function to be a bit clearer in the prose that it refers specifically to unbound symbols in `QkParam` instances?

Cryoris replied:
Good idea, I'll update that 👍🏻

Cryoris replied:
Done in 7abe7bc

jakelishman commented on crates/cext/src/circuit.rs:
```diff
@@ -1079,7 +1205,10 @@ pub unsafe extern "C" fn qk_circuit_instruction_clear(inst: *mut CInstruction) {
         inst.num_clbits = 0;
         if inst.num_params > 0 && !inst.params.is_null() {
             let params = std::slice::from_raw_parts_mut(inst.params, inst.num_params as usize);
-            let _ = Box::from_raw(params as *mut [f64]);
+            for param in params.iter() {
+                let _ = Box::from_raw(*param);
+            }
+            let _ = Box::from_raw(params);
```
Woah boy this `unsafe` block is _miles_ too big.  I was going to comment "where's the `unsafe`?" for `Box::from_raw` and then I realise the entire function is wrapped in it...

Imo, ideally _every_ `unsafe` function call has its own `unsafe` block so the reasoning is enforced to be local.  `unsafe` _should_ be annoying and wordy to use - it's really dangerous - and it _should_ motivate us to encapsulate things in safer interfaces where possible.

Cryoris replied:
I can do this in a follow up for all calls here

jakelishman commented on crates/cext/src/circuit.rs:
```diff
@@ -1079,7 +1205,10 @@ pub unsafe extern "C" fn qk_circuit_instruction_clear(inst: *mut CInstruction) {
         inst.num_clbits = 0;
         if inst.num_params > 0 && !inst.params.is_null() {
             let params = std::slice::from_raw_parts_mut(inst.params, inst.num_params as usize);
-            let _ = Box::from_raw(params as *mut [f64]);
+            for param in params.iter() {
+                let _ = Box::from_raw(*param);
+            }
+            let _ = Box::from_raw(params);
```
This looks correct to me, but makes me wonder if it would be easier for us to allocate the params like
```rust
let params = mem::ManuallyDrop::new(
    params_view().iter().cloned().collect::<Box<[Param]>>()
);
let params_ptr = mem::ManuallyDrop::new(
    params.iter_mut().map(std::ptr::from_mut).collect::<Box<[*mut Param]>>()
);
CInstruction {
    params: params_ptr.as_mut_ptr(),
    num_params: params_ptr.len(),
    ..
}
```
on creation, then on drop:
```rust
let inst: CInstruction;
{
    let params = boxed_slice_from_ptr(inst.params, inst.num_params);
    let _ = boxed_slice_from_ptr(params[0], inst.num_params);
}
```
or something like that, rather than new individual heap allocations for each pointer?

There'd be an advantage of memory locality, fewer calls to the allocator, and the drop code ends up a little less manual.

The disadvantage is that even though the `CInstruction` is documented as having been copy-out, you still mustn't mutate pointed-to elements of `params` or we'd have de-allocation bugs (and we'd need to document that). (Tbh though, we should document that `CInstruction` itself isn't mutable.)

I feel on the fence here - do you have a preference?

It's a shame we can't directly store `params: *mut Param` in `CInstruction` because C doesn't know the size of the `Param` type to be able to do pointer arithmetic.

Cryoris replied:
Oh I didn't know `mem::ManuallyDrop` that's pretty neat -- we can just use that instead of `Box::into_raw` when we're newly creating the data? 

> The disadvantage is that even though the CInstruction is documented as having been copy-out, you still mustn't mutate pointed-to elements of params or we'd have de-allocation bugs (and we'd need to document that).

This is potentially a pretty nasty edge case, I think I'd be in favor of a more stable setup like we have now. I could run a small benchmark to see if your suggestion is faster, if you think it would be noticeable?

jakelishman replied:
Let's just leave it as-is for now - it's probably not worth the effort for something we will (hopefully) re-do in the future anyway, not to mention the sharp edges.

jakelishman commented on crates/cext/src/circuit.rs:
```diff
@@ -456,6 +496,89 @@ pub unsafe extern "C" fn qk_circuit_gate(
+            // There are no ``QkGate``s > 4 qubits
+            _ => unreachable!(),
+        }
+    };
```
This is a pretty funky way of spelling something that's basically just `std::slice::from_raw_parts(qubits, gate.num_qubits())` with a subsequent slice cast.

I think it's pre-existing though, so let's leave it for now if so.  It's crying out for a helper function though (and tbh, I feel like I've written a `bytemuck` cast for `&[Qubit] -> &[PhysicalQubit]` somewhere, so the job may already be done).

Cryoris replied:
yeah it didn't feel natural to even copy haha, let's add a more automated way for both append functions in a follow up

jakelishman commented on crates/cext/src/circuit.rs:
```diff
@@ -338,6 +338,46 @@ pub unsafe extern "C" fn qk_circuit_num_clbits(circuit: *const CircuitData) -> u
+/// QkParam *y = qk_param_new_symbol("y");
+///
+/// uint32_t q0[1] = {0};
+/// const QkParam *rx_param[1] = {x};
+/// const QkParam *ry_param[1] = {y};
+///
+/// qk_circuit_parameterized_gate(qc, QkGate_RX, q0, rx_param);
+/// qk_circuit_parameterized_gate(qc, QkGate_RY, q0, ry_param);
```
It doesn't matter for your example, but technically
```c
qk_circuit_parameterized_gate(qc, QkGate_RX, q0, &x);
qk_circuit_parameterized_gate(qc, QkGate_RY, q0, &y);
```
is also fine (and tbh, likely identical to what the compiler will generate anyway).

jakelishman commented on crates/cext/src/circuit.rs:
```diff
@@ -456,6 +496,89 @@ pub unsafe extern "C" fn qk_circuit_gate(
+/// @param params The pointer to the array of ``QkParam*`` parameters to use for the gate parameters.
+/// This can be a null pointer if there are no parameters for ``gate`` (e.g. ``QkGate_H``).
+///
+/// @return An exit code.
```
It'd be good to document what the detected error mechanisms actually are, so C knows.

Cryoris replied:
Done in 7abe7bc

jakelishman commented on crates/cext/src/circuit.rs:
```diff
@@ -456,6 +496,89 @@ pub unsafe extern "C" fn qk_circuit_gate(
+    match circuit.push_standard_gate(gate, &params, qargs) {
+        Ok(_) => ExitCode::Success,
+        Err(_) => ExitCode::ParameterError,
+    }
```
`push_standard_gate` actually returns a semi-structured error, so we _should_ be able to match on it and make sure we're returning the proper code and we don't silently return nonsense if the signature/errors from it change in the future.

Alternatively (and/or maybe in a future PR) we could make an
```rust
impl From<CircuitDataError> for ExitCode {}
impl From<Result<(), CircuitDataError>> for ExitCode {}
```
and then use that in returns.

(Also we need to split up `CircuitDataError` to get more complete type safety out of it, but this is still a partial step we can take.)

Cryoris replied:
I think this PR might predate `CircuitDataError`, let me see if I can nicely add this

Cryoris replied:
Ok a generic `From<CircuitData> for ExitCode` is beyond the scope of this PR but I added a case to extract a name conflict, which is the only actionable error for the user here

jakelishman commented on releasenotes/notes/qk-gate-param-e86fb60fa75bd7e8.yaml:
```diff
@@ -0,0 +1,24 @@
+features_c:
+  - |
+    Added support to add parameterized gates to circuits using the new
+    ``qk_circuit_gate_param`` function, which specifies the gate parameters as
```
```suggestion
    :c:func:`qk_circuit_gate_param` function, which specifies the gate parameters as
```
function cross refs should work within the `c` domain

jakelishman commented on releasenotes/notes/qk-gate-param-e86fb60fa75bd7e8.yaml:
```diff
@@ -0,0 +1,24 @@
+      qk_circuit_gate_param(qc, QkGate_RX, qubit, params); // add RX(theta) to the circuit
+
+  - |
+    Added ``qk_circuit_num_symbols`` to query the number of unbound symbols present
```
```suggestion
    Added :c:func:`qk_circuit_num_symbols` to query the number of unbound symbols present
```

jakelishman commented on releasenotes/notes/qk-gate-param-e86fb60fa75bd7e8.yaml:
```diff
@@ -0,0 +1,24 @@
+    in the circuit.
+upgrade_c:
+  - |
+    ``QkInstruction.params`` was upgraded from a pointer to ``double`` to a pointer
```
```suggestion
    :c:member:`QkInstruction.params` was changed from a pointer to ``double`` to a pointer
```
I think `:c:member:` is the right cross-ref role, and "changed" is perhaps more correct than "upgraded".  The "upgrade" section is named that because it's what users have to take care of on a version upgrade.

jakelishman commented on test/c/test_circuit.c:
```diff
@@ -979,6 +980,164 @@ static int test_not_unitary_gate(void) {
+    if (!qk_param_equal(angle, theta)) {
+        result = EqualityError;
+    }
+    qk_param_free(angle);
```
This code isn't valid per documentation of `CInstruction` - you _mustn't_ free the pointer yourself.  The test only gets away with it because it also forgets to free `inst`.

Cryoris replied:
Whoops! Fixing it!

Cryoris replied:
Done in 7abe7bc

jakelishman reviewed (commented):
In principle I'm fine with this to move forwards.  You know my reservations about `QkParam` and the Rust-space `Param` object in general, but re-iterating those isn't going to help us make progress.

Cryoris committed:
Apply suggestions from code review

Co-authored-by: Jake Lishman <jake@binhbar.com>

Cryoris committed:
Jake's review comments

Cryoris committed:
Merge branch 'main' into qk-gate-param

Cryoris committed:
Fix circuit comparison

Cryoris committed:
Merge branch 'qk-gate-param' of github.com:Cryoris/qiskit-terra into qk-gate-param

jakelishman committed:
Merge branch 'main' into qk-gate-param

jakelishman commented on crates/cext/src/circuit.rs:
```diff
@@ -457,6 +498,102 @@ pub unsafe extern "C" fn qk_circuit_gate(
+            // conflict, so we match on that (until a more generic CircuitDataError -> QkExitCode
+            // is implemented)
+            if let CircuitDataError::ParameterTableError(ParameterTableError::NameConflict(_)) =
+                circuit_error
+            {
+                return ExitCode::ParameterNameConflict;
+            }
+            ExitCode::ParameterError
+        }
+    }
```
No need to change the PR for just this, but this is a pretty awkward spelling.  Something like

```suggestion
    match circuit.push_standard_gate(gate, &params, qargs) {
        Ok(()) => ExitCode::Success,
        Err(CircuitDataError::ParameterTableError(ParameterTableError::NameConflict(_))) => {
            ExitCode::ParameterNameConflict
        }
        Err(_) => ExitCode::ParameterError,
    }

```

is perhaps more natural.

jakelishman commented on releasenotes/notes/qk-gate-param-e86fb60fa75bd7e8.yaml:
```diff
@@ -0,0 +1,24 @@
+      QkParam *theta = qk_param_new_symbol("theta");
+      uint32_t qubit[1] = {0};
+      const QkParam* params[1] = {theta};
+      qk_circuit_gate_param(qc, QkGate_RX, qubit, params); // add RX(theta) to the circuit
```
```suggestion
      qk_circuit_parameterized_gate(qc, QkGate_RX, qubit, params); // add RX(theta) to the circuit
```

jakelishman commented on test/c/test_circuit.c:
```diff
@@ -979,6 +980,164 @@ static int test_not_unitary_gate(void) {
+        goto cleanup;
+    }
+
+    // Free the circuit and check the parameter is still correctly accessible
+    // from the instruction itself
+    qk_circuit_free(qc);
+
+    QkParam *angle = inst.params[0];
+    if (!qk_param_equal(angle, theta)) {
+        result = EqualityError;
+    }
```
We'd probably really have to run within `valgrind` or some other memory sanitiser for this test to really do its full job reliably, but I'm fine leaving it as is.

jakelishman reviewed (commented):
Thanks for this Julien.  This is ready to go other than the typo in the code example in the release note.

(That said, I'm not 100% confident in my own reviewing right now - feeling a bit drained coming up on the release!)

Cryoris committed:
Apply suggestions from code review

Co-authored-by: Jake Lishman <jake@binhbar.com>

Cryoris committed:
Satisfy clippy
